### PR TITLE
Migration/tuplet

### PR DESCRIPTION
**Pull request by rvilarl**
Tuplet!

### CODE REVIEW

File: src/note.ts [hunk 1]
```diff
@@ -498,7 +498,15 @@ export abstract class Note extends Tickable {
   getNoteType(): string {
     return this.noteType;
   }
-
+  getBeam(): Beam | undefined {
+    return this.beam;
+  }
```
Comment (0xfe) [inline src/note.ts:503]: New line after paren please.

File: src/beam.ts [hunk 2]
```diff
@@ -411,14 +411,15 @@ export class Beam extends Element {
...
+      const direction =
+        (tuplet.notes[0] as StemmableNote).stem_direction === Stem.DOWN ? Tuplet.LOCATION_BOTTOM : Tuplet.LOCATION_TOP;
       tuplet.setTupletLocation(direction);
 
       // If any of the notes in the tuplet are not beamed, draw a bracket.
       let bracketed = false;
       for (let i = 0; i < tuplet.notes.length; i++) {
         const note = tuplet.notes[i];
-        if (note.beam === undefined) {
+        if (note.getBeam() === undefined) {
```
Comment (0xfe) [inline src/beam.ts:422]: Can we define a `hasBeam` in `note.ts` similar to `hasStem`?

File: src/tuplet.ts [hunk 3]
```diff
@@ -81,7 +110,7 @@ export class Tuplet extends Element {
     if ('bracketed' in this.options) {
       this.bracketed = this.options.bracketed;
     } else {
-      this.bracketed = notes.some((note) => note.beam === undefined);
+      this.bracketed = notes.some((note) => note.getBeam() === undefined);
```
Comment (0xfe) [inline src/tuplet.ts:113]: You can use `hasBeam` as mentioned above.

File: src/tuplet.ts [hunk 4]
```diff
@@ -310,17 +339,12 @@ export class Tuplet extends Element {
...
-          this.location * 10
-        );
-        this.context.fillRect(
+        ctx.fillRect(this.x_pos, this.y_pos, line_width, 1);
+        ctx.fillRect(this.x_pos + this.width / 2 + width / 2 + 5, this.y_pos, line_width, 1);
+        ctx.fillRect(this.x_pos, this.y_pos + +(this.location === Tuplet.LOCATION_BOTTOM), 1, this.location * 10);
+        ctx.fillRect(
           this.x_pos + this.width,
-          this.y_pos + (this.location === Tuplet.LOCATION_BOTTOM),
+          this.y_pos + +(this.location === Tuplet.LOCATION_BOTTOM),
```
Comment (0xfe) [inline src/tuplet.ts:347]: I'm not sure happening here?

**Review by 0xfe — COMMENTED** [review]
Overall looks good. Some minor comments.

File: src/tuplet.ts (hunk 4 above)
Comment (rvilarl) [reply src/tuplet.ts:347]: What was the intention in origin? I used the solution from @wassertim I also quite not get it summing a number and a boolean.

File: src/note.ts (hunk 1 above)
Comment (rvilarl) [reply src/note.ts:503]: done

File: src/beam.ts (hunk 2 above)
Comment (rvilarl) [reply src/beam.ts:422]: Done

File: src/tuplet.ts (hunk 3 above)
Comment (rvilarl) [reply src/tuplet.ts:113]: done

File: src/tuplet.ts (hunk 4 above)
Comment (rvilarl) [reply src/tuplet.ts:347]: I have changed it to a ?: construction. Easie to understand.

**Comment by rvilarl** [comment]
Ready to merge!

File: src/tuplet.ts (hunk 4 above)
Comment (0xfe) [reply src/tuplet.ts]: Yeah, that's better -- not sure how the original version got that, definitely not readable.

**Comment by 0xfe** [comment]
Looks good.

**Pull request merged**